File: leetcodeDailychallenge/2095. Delete the Middle Node of a Linked List.java
//problem link: https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/?envType=daily-question&envId=2026-06-15
//timeComplexity: O(n/2)
//spaceComplexity: O(1)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteMiddle(ListNode head) {
        if(head.next==null){
            return null;
        }
        ListNode slow=head;
        ListNode fast=head;
        ListNode last=null;
        while(fast!=null && fast.next!=null){
            fast=fast.next.next;
            last=slow;
            slow=slow.next;
        }
        last.next=slow.next;
        slow.next=null;
        return head;
    }
}

File: LINKED-LISTS/2095. Delete the Middle Node of a Linked List.cpp
//problem link: https://leetcode.com/problems/delete-the-middle-node-of-a-linked-list/
//timeComplexity: O(n)
//spaceComplexity: o(1)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteMiddle(ListNode* head) {
        if(head->next==NULL){
            return NULL;
        }
        ListNode*slow=head;
        ListNode*fast=head;
        ListNode*prevNode=NULL;
        while(fast!=NULL && fast->next!=NULL){
            prevNode=slow;
            slow=slow->next;
            fast=fast->next->next;
        }
        prevNode->next=slow->next;
        slow->next=NULL;
        return head;
    }
};